doc/dev/design/FABRID.rst

```diff
@@ -0,0 +1,605 @@
+This results in a set of *PolicyIdentifiers* per hop, which can then be used by the application, such as by defining an application parameter
+that then selects the policies to use on the path and hands these to the snet implementation, e.g. ``--fabridpolicy``. 
+
+
+``fabridpolicy`` parameter
+''''''''''''''''''''''''''''
+A custom language is used to make a selection out of the available paths and policies. The basic components of the language are as follows:
+
+* **Identifiers**
+
+  An identifier matches with a specific hop in the path and applies a policy to that hop. Parts of this hop identifier may be a wildcard,
+  such that the identifier can match with multiple hops in the path.
+  An identifier is structured as follows: 
+  ``ISD-AS#IGIF,EGIF@POLICY``,
+  where 
+
+  * ISD can be either the ISD number (e.g. ``1``), or a wildcard (``0``).
+  * AS can be either the AS number seperated by underscores (e.g. ``ff00_0_110``) or a wildcard (``0``).
+  * IGIF can be either the ingress interface number (e.g. ``42``), or a wildcard (``0``).
+  * EGIF can be either the egress interface number (e.g. ``41``), or a wildcard (``0``).
+  * POLICY can be either the policy to apply, where a local policy is denoted as ``L`` + the policy identifier (e.g. ``L100``) and a global policy
+    is denoted by ``G`` + the policy identifier (e.g. ``G100``), a wildcard (``0``), or a rejection ``REJECT``.
+    Rejection means that this path should not be chosen. 
+
+  When used in a query, the identifier evaluates to true when at least a single hop in the path matches the identifier.
+
+* **Concatenations**
+
+  Multiple identifiers can be combined by using a concatenation (and/or parentheses). Concatenations are created by the ``+`` symbol. 
+
+  Example:
+
+  ``(1-0#0,0@G300 + 1-0#0,0@G200)`` applies both policy G300 and policy G200.
+
+  When used in a query, a concatenation evaluates to true when *all* identifiers in the concatenation also evaluate to true.
+
+
+* **Queries**
+
+  You can query for the existence of a specific hop and/or policy through a query.
+  Queries are structured as follows: ``{ QUERY_EXPRESSION ? EXPRESSION_IF_TRUE : EXPRESSION_IF_FALSE}``.
+  The query expression is evaluated, and if an identifier matches with a specific hop, the ``expression_if_true`` branch is applied.
+  If no matches can be found in the path, the ``expression_if_false`` branch is applied.
+  Identifiers in the query expression are not applied, e.g. if a query expression queries for a specific policy, the specific policy
+  is not applied to the hops it matches, unless the same expression is also given under the ``expression_if_true`` branch.
+
+  Example:
+
+  With the path
+  ``1-ff00:0:109#0,5@() 1-ff00:0:110#4,1@(G100, G200) -> 1-ff00:0:111#2,0@(G200, G300)``
+
+  When an expression queries for ``1-0#0,0@G200`` using ``{1-0#0,0@G200 ? 1-0#0,0@G300 : 1-0#0,0@REJECT}``, the policies that are
+  applied to the hops are only policy G300 for the last hop.
```

Comment:
Please explain why / when it makes sense to apply a policy different from the policy that is queried.
Also, would be nice to explicitly explain how (some of) the use cases in the introduction can now be achieved.
E.g., how we strictly enforce traffic to only be routed along devices of manufacturer X? (And if an AS does not support FABRID, or does not have a intra-AS path with devices only from manufacturer X, then we want to avoid that AS)

Reply:
@jeltevanbommel

Reply:
Queries were mostly for conditional rejection of the paths. For context: there are two ways to reject a path based on policy: 
- a whitelist (a concatenation of allowed policies and at the end of the concatenation a wildcard REJECT policy)
- a blacklist (if a specific policy occurs, reject the paths)
I imagine the whitelist will get more use, but the blacklist mechanism may also be necessary in some cases.  I have added an example below. I can also imagine that this query mechanism is extended in the future, to only apply policies from EXPRESSION_IF_TRUE on the matches from the QUERY_EXPRESSION.